yourbench/pipeline/chunking.py

```diff
@@ -0,0 +1,608 @@
+        if consecutive_sims:
+            all_similarities.append(consecutive_sims)
+
+        # Create single-hop chunks
+        single_hop_chunks = _chunk_document(
+            sentences=sentences,
+            similarities=consecutive_sims,
+            l_min_tokens=l_min_tokens,
+            l_max_tokens=l_max_tokens,
+            tau=tau_threshold,
+            doc_id=doc_id,
+        )
+
+        # Create multi-hop chunks (modified to ensure no duplicates)
+        multihop = _multihop_chunking(
+            single_hop_chunks, h_min=h_min, h_max=h_max, num_multihops_factor=num_multihops_factor
+        )
+
+        # Compute metrics (token_count, perplexity, readability, etc.)
+        chunk_metrics = _compute_info_density_metrics(single_hop_chunks, local_perplexity_metric, local_use_textstat)
+
+        # Accumulate
+        all_single_hop_chunks.append(single_hop_chunks)
+        all_multihop_chunks.append(multihop)
+        all_chunk_info_metrics.append(chunk_metrics)
+
+    # Optional: Save aggregated similarity plot
+    if all_similarities and debug_mode:
+        _plot_aggregated_similarities(all_similarities)
+
+    # Append new columns to dataset
+    dataset = dataset.add_column("chunks", all_single_hop_chunks)
+    dataset = dataset.add_column("multihop_chunks", all_multihop_chunks)
+    dataset = dataset.add_column("chunk_info_metrics", all_chunk_info_metrics)
+    dataset = dataset.add_column("chunking_model", [model_name] * len(dataset))
+
+    # Save updated dataset
+    try:
+        save_dataset(dataset, "chunking", config, output_dataset_name, output_subset)
+        logger.success(
+            "Chunking stage complete. Dataset saved to '{}', subset '{}'.", output_dataset_name, output_subset
+        )
+    except Exception as save_error:
+        logger.error("Failed to save chunked dataset for doc '{}': {}", output_dataset_name, save_error)
+
+
+def _split_into_sentences(text: str) -> List[str]:
+    """
+    Splits the input text into sentences using a simple rule-based approach
+    that looks for punctuation delimiters ('.', '!', '?').
+
+    Args:
+        text (str): The full document text to be split.
+
+    Returns:
+        List[str]: A list of sentence strings.
+    """
+    # Replace newlines with spaces for consistency
+    normalized_text = text.replace("\n", " ").strip()
+    if not normalized_text:
+        return []
+
+    # Split using capturing parentheses to retain delimiters, then recombine.
+    segments = re.split(r"([.!?])", normalized_text)
+    sentences: List[str] = []
+    for i in range(0, len(segments), 2):
+        if i + 1 < len(segments):
+            # Combine the text and delimiter
+            candidate = (segments[i] + segments[i + 1]).strip()
+        else:
+            # If no delimiter segment, use the text directly
+            candidate = segments[i].strip()
+        if candidate:
+            sentences.append(candidate)
+    return sentences
+
+
+def _compute_embeddings(
```

Comment:
this works well - but i want to be careful and make sure that this is very efficient (as this is the one thing that we'll locally need hardware for on the hf-space)

Reply:
@alozowski has worked on something there so she might want to integrate her changes here